Document and implement doubly-linked implementation of prop (see #1040)

blefloch opened the pull request:
There remains \prop_show:N and log, which means that we cannot really
test the code seriously yet.  However, it is already exercised as
part of the implementation of \prop_concat:NNN and the from_keyval
functions.

josephwright commented:
Any news on this: I think 'large' props would be a great addition (cf. work on tagging for example).

car222222 commented:
What about "props as hash-table" (since this seems to be in use nearly in the kernel by now)??

blefloch commented:
I decided to rebase locally on the `main` branch then force-pushed rather than having a pretty painful merge.  I also squashed the commits into a single one because there was no natural way to split up the code changes.  For now the "added" date is today 2024-01-15 to make l3doc happy.  At long last, this pull request is ready to be reviewed and merged, I think.

Skillmon commented on l3kernel/l3prop.dtx:
```diff
@@ -49,32 +49,73 @@
+%     \cs[no-index]{prop_map_\ldots{}}.
+%
+%   \item
+%     The \enquote{linked} storage method is meant for property lists with a
```
```suggestion
%     The \enquote{linked} storage method is meant for property lists with
```

josephwright reviewed (approved):
Looks good to me at the code level - lets see what everyone thinks interface-wise.

FrankMittelbach commented on l3kernel/l3prop.dtx:
```diff
@@ -148,17 +200,33 @@
-%   \pkg{l3keys}), each key here \emph{must} be followed with an \texttt{=}
-%   sign.
+% \begin{function}[added = 2024-01-15]
+%   {\prop_const_large_from_keyval:Nn, \prop_const_large_from_keyval:cn}
```
again "linked" feels more natural to me

FrankMittelbach reviewed (commented):
Looks good to me, but I'm not through reviewing. Any, here are some first comments on the interface section

blefloch commented on l3kernel/l3prop.dtx:
```diff
@@ -49,32 +49,73 @@
+%
+% \begin{function}[added = 2024-01-15]{\prop_new_large:N, \prop_new_large:c}
+%   \begin{syntax}
+%     \cs{prop_new_large:N} \meta{property list}
```
You correctly pinpointed my reason to prefer "large" in function names (and my inconsistency in using a different word in error messages), which is that linked is an implementation detail.  In fact, the new props are both doubly-linked and implemented using the hash table, and the latter point is an important part of why they end up faster for accessing entries.  Is "linked" really a good word for that? "hashed"?

car222222 replied:
I agree that names that refer to the implementation are not good (since there may be further changes to this in another decade or so). 

I also understand @FrankMittelbach ‘s problem with labelling them as “large”.

Needs further thought!

car222222 replied:
Further: in some of the documentation, it is probably best to use both names.

car222222 replied:
We need a name that does not imply any “ordering” on the type of implementation, and says nothing about the expected contents, but is also independent of any details of the implementation.

blefloch commented on l3kernel/l3prop.dtx:
```diff
@@ -49,32 +49,73 @@
+% using \cs{prop_new:N} (\enquote{flat} storage) or \cs{prop_new_large:N}
+% (\enquote{linked} storage). Once a property list is declared with
+% \cs{prop_new:N} or \cs{prop_new_large:N}, the type of internal data storage
+% can no longer be changed. All other \pkg{l3prop} functions transparently
```
Easy to provide but (1) what name? (2) for local props it's probably annoying if people try to change the storage type within a group, I'll have to see how much bookkeeping is needed.

car222222 replied:
I would agree that changing this “type” only locally does not sound like a good idea. 

Name: \prop_make_large:N

car222222 replied:
Note that \prop_set_eq:NN will do the conversion if needed (both ways). 
So it just needs a small extension to cope with the name staying the same??

blefloch replied:
`\prop_make_...:N` is a good name.  I also thought of `\prop_renew_...:N` but it might suggest that the process would give an empty prop like `\prop_new_...:N`.  Regardless of name we need both directions I suppose.

There is a bit more work than for `\prop_set_eq:NN` because some auxiliary structures have to be added / deleted.

car222222 commented:
The diff here is "too big" for my github app!  It refuses to load it.

car222222 commented on l3kernel/l3prop.dtx:
```diff
@@ -604,27 +677,31 @@
+%   and specifies its storage type.
 % \end{function}
 %
 % \section{Scratch property lists}
```
It is probably sensible to add a note here that the are no scratch “large” pls.

car222222 commented:
But I can get the diff using Chrome.  

But response is achingly slow: not sure if this is due to the size of the diff or just my connection? 

So I have now added a few responses and comments on the documentation.  Not yet tackled the code.

FrankMittelbach commented:
A suggested stress test for the implementation: 

force all property lists to be"large" in the test suite by making `\prop_new:N` equal to `\prop_new_large:N` in the kernel and then run the 2e test suite. It would be very interesting to see what that would do to the time it takes to run the tests in all directories (compared to the time it currently needs).

blefloch commented:
Indeed Frank that would be a good test, but I'll need some time to set it up (pretty busy work week).  I'm worried because the order of prop items is probably different with the new prop implementation than the old one, which may show up in various test logs.

car222222 commented:
In the explanation of the prefix, add something like this:

The aim here is to make this string as short as we can, given the range of distinct characters available. 

(Maybe also explain why it should be as short as possible, possibly by reference to earlier where the reason is fully explained?)

u-fischer commented on l3kernel/l3prop.dtx:
```diff
@@ -49,32 +49,73 @@
+%
+%   \item
+%     The \enquote{linked} storage method is meant for property lists with a
+%     large numbers of entries.  It has more memory overhead, but is
```
Imho "memory overhead" should be clarified. Half of the user will probably think that that means computer memory.

car222222 commented:
At a later stage, the code for the production of short, unique strings should probably be moved to somewhere that makes it more widely available.

blefloch commented:
I think I addressed all of the comments.  I also found a few bugs in my code, and minor ones elsewhere (see recent issues).

Most importantly, I renamed the functions as Frank suggested, `\prop_new_linked:N` etc., and I had to spend quite some time fighting with the scope checking for l3debug.  From my point of view this could be merged.

The small merge conflict with main is trivial to fix: keep my additional `\__prop_chk:w` and remove the line break.  The failing tests seem related to Joseph's l3build changes.  Rebasing on main should fix them I suppose.

josephwright commented:
Looks good to me: lets see what everyone else thinks!

blefloch commented:
Ah, one check that Frank suggested was to run the LaTeX2e suite with an additional line making all props linked.  I didn't figure out how to do that, but in case someone wants to try, one has to add `\cs_gset_eq:NN \prop_new:N \prop_new_linked:N` after the definition of `\prop_new_linked:N` in `l3prop.dtx`, and to comment out `\str_if_eq:nnF {#3} { flat } { #3~ }` in `l3msg.dtx`.  This makes a failing `m3prop007` file because of some missing `linked ` words, but it will minimize the number of test log changes on the LaTeX2e side.

blefloch commented:
I managed to use the new linked prop in the latex2e test suite, which uncovers a problem with treatment of hashes.  I'm investigating.

I also forgot to add the suggested `\prop_make_linked:N` and `\prop_make_flat:N` functions (and global versions `\prop_gmake_linked:N` and `\prop_gmake_flat:N` I suppose).

josephwright commented:
Presumably we can apply the 'make flat' function in 2e rollback if available: either the `prop` is flat and 'make flat' is unvailable, or the 'make flat' function is available and will work.

blefloch commented:
Yes.  Anyways, props are flat by default, so this branch doesn't change anything in LaTeX2e.  I just wanted to test whether the LaTeX2e test suite runs correctly with props being linked props.  Overall it slows down the test suite slightly there, presumably because there are not that many very long props.

Despite the failing tests, I think the branch can be merged/rebased into main.  Last I checked (before updating l3build with the latest normalizations), the LaTeX3 tests succeeded.
